src/layers/basic.jl

```diff
@@ -503,3 +503,61 @@ function Base.show(io::IO, e::Embedding)
 end
 
 outputsize(e::Embedding) = (e.out_dims,)
+
+"""
+    PeriodicEmbedding(dims, periods)
+
+Create an embedding periodic in some dimensions with specified periods. Dimensions not in
+`dims` are passed through unchanged, but dimensions in `dims` are moved to the end of the
+output and replaced with their sines, followed by their cosines (scaled appropriately to
+have the specified periods). This smooth embedding preserves phase information and enforces
+periodicity.
+
+## Arguments
+
+  - `dims`: Periodic dimensions
+  - `periods`: Periods of the dimensions `dims`, in the same order
+
+## Inputs
+
+  - `x` must be an `AbstractArray` with `issubset(dims, axes(x, 1))`
+  - `st` must be a `NamedTuple` where `st[:k] = 2 ./ periods`, but is on the same device as
+    `x`
+
+## Returns
+
+  - `AbstractArray` of size `(size(x, 1) + length(dims), ...)` where `...` are the other
+    dimensions of `x`.
+  - `NamedTuple` with field `k` equal to `st[:k]`, but with elements of the same type as the
+    elements of `x`
+"""
+@concrete struct PeriodicEmbedding <:AbstractExplicitLayer
+    dims
+    periods
+end
+
+Lux.initialstates(::AbstractRNG, p::PeriodicEmbedding) = (k = 2 ./ p.periods,)
+
+@inline function (p::PeriodicEmbedding)(x::AbstractVector, ps, st::NamedTuple)
+    return vec(first(p(reshape(x, :, 1), ps, st))), st
+end
+
+@inline function (p::PeriodicEmbedding)(x::AbstractMatrix{T}, ps, st::NamedTuple) where T
+    other_dims = ChainRulesCore.@ignore_derivatives setdiff(axes(x, 1), p.dims)
+    _st = (k = T.(st[:k]), )
```

Comment:
I'm not sure we actually want to have this casting here, but it seems necessary to match the type of the output and `x`. Is there a better way to approach this? Should the type that people originally used when specifying `periods` matter?

Reply:
Generally, I prefer to just use anything that is supplied and let julia semantics determine what happens. The main issue with these sort of type-casting (except in situations where without casting you hit an error) is we don't have a way to turn them off from outside.

Reply:
Okay, I've removed the type-casting. I was mostly trying to address your earlier comment about accidentally type-casting everything to Float64.

Reply:
Ah right, I was referring to using $\pi$ directly which forces a conversion, here users are kind of asking for it but supplying a different eltype in the states